### PR TITLE
Fix wait4 failure of child process created with vfork

### PR DESCRIPTION
**Pull request by jessehui**

### CODE REVIEW

File: src/libos/src/process/do_wait4.rs [hunk 1]
```diff
@@ -54,7 +58,8 @@ pub fn do_wait4(child_filter: &ProcessFilter, options: WaitOptions) -> Result<(p
         .collect::<Vec<&ProcessRef>>();
 
     if unwaited_children.len() == 0 {
-        return_errno!(ECHILD, "Cannot find any unwaited children");
+        // Check again for vfork-ed child process which doesn't have a real structure of a process
+        return wait4_exit_child_created_with_vfork(process.pid(), child_filter);
```
Comment (volcano0dr) [inline src/libos/src/process/do_wait4.rs:62]: If there are child processes and all child processes are running, the vfork-ed child process of the terminated should also be checked.

File: src/libos/src/process/do_wait4.rs [hunk 2]
```diff
@@ -30,6 +31,9 @@ pub fn idle_reap_zombie_children() -> Result<()> {
         }
     }
 
+    // Also check for vforked child process
+    reap_zombie_child_created_with_vfork();
```
Comment (volcano0dr) [inline src/libos/src/process/do_wait4.rs:35]: I don't think this function should be called here, this function should be used to clean up memory when libos is shutdown.

File: src/libos/src/process/do_wait4.rs (hunk 1 above)
Comment (jessehui) [reply src/libos/src/process/do_wait4.rs]: Fixed. Thanks for your review.

File: src/libos/src/process/do_wait4.rs (hunk 2 above)
Comment (jessehui) [reply src/libos/src/process/do_wait4.rs]: Yes. This is wrong. I re-implement this function and call it in `exit_process`. Please take a new look. Thank you for your review.

File: src/libos/src/process/do_vfork.rs [hunk 3]
```diff
@@ -201,3 +228,62 @@ pub fn handle_force_stop() {
         }
     }
 }
+
+// Wait4 unwaited child which are created with vfork and directly exit without calling execve.
+pub fn wait4_exit_child_created_with_vfork(
+    parent_pid: pid_t,
+    child_filter: &ProcessFilter,
+) -> Result<(pid_t, i32)> {
```
Comment (volcano0dr) [inline src/libos/src/process/do_vfork.rs:236]: I think it makes more sense for the return type of `wait4_exit_child_created_with_vfork` to be `Option<(pid_t, i32)>`,  the error of no child process is handled in `do_wait4`.

File: src/libos/src/process/do_vfork.rs [hunk 4]
```diff
@@ -201,3 +228,62 @@ pub fn handle_force_stop() {
...
+            }
+            return Ok((*child.pid(), child.status().as_u32() as i32));
+        }
+    }
+
+    return_errno!(ECHILD, "no exit child found");
+}
+
+// Reap all unwaited child which are created with vfork and directly exit without calling execve.
+pub fn reap_zombie_child_created_with_vfork(parent_pid: pid_t) -> Vec<pid_t> {
```
Comment (volcano0dr) [inline src/libos/src/process/do_vfork.rs:258]: reap_zombie_child_created_with_vfork(parent_pid: pid_t) -> Option<Vec<pid_t>>

File: src/libos/src/process/do_vfork.rs (hunk 3 above)
Comment (jessehui) [reply src/libos/src/process/do_vfork.rs]: Done. Thanks for your review.

File: src/libos/src/process/do_vfork.rs (hunk 4 above)
Comment (jessehui) [reply src/libos/src/process/do_vfork.rs]: Done. Thanks for your review.

**Review by volcano0dr — APPROVED** [review]
LGTM

**Pull request merged**